DistroLauncher-Tests/DistroLauncher-Tests/named_mutex_tests.cpp

```diff
@@ -0,0 +1,283 @@
+    {
+        return static_cast<dummy_mutex*>(mutex_handle)->locked;
+    }
+};
+
+class TestNamedMutex : public NamedMutexWrapper<TestMutexApi>
+{
+  public:
+    TestNamedMutex(std::wstring name, bool lazy_init = false) : NamedMutexWrapper(name, lazy_init)
+    { }
+
+    // Exposing internal state for testing
+    HANDLE& get_mutex_handle()
+    {
+        return mutex_handle;
+    }
+    bool& locked()
+    {
+        return TestMutexApi::locked(mutex_handle);
+    }
+};
+
+std::list<TestMutexApi::dummy_mutex> TestMutexApi::dummy_back_end{};
+
+// Overriding back-end API
+DWORD TestMutexApi::create(HANDLE& mutex_handle, LPCWSTR mutex_name) noexcept
+{
+    auto it = std::find(dummy_back_end.begin(), dummy_back_end.end(), mutex_name);
+    if (it == dummy_back_end.cend()) {
+        dummy_back_end.push_back(dummy_mutex{std::wstring(mutex_name)});
+        mutex_handle = static_cast<HANDLE>(&dummy_back_end.back());
+    } else {
+        ++it->refcount;
+        mutex_handle = static_cast<HANDLE>(&(*it));
+    }
+    return 0;
+}
+
+DWORD TestMutexApi::destroy(HANDLE& mutex_handle, LPCWSTR mutex_name) noexcept
+{
+    auto it = std::find(dummy_back_end.begin(), dummy_back_end.end(), mutex_name);
+
+    if (it == dummy_back_end.cend()) {
+        return 1; // Destroyed a non-existing mutex
+    }
+
+    --it->refcount;
+    if (it->refcount == 0) {
+        dummy_back_end.erase(it);
+    }
+    return 0;
+}
+
+DWORD TestMutexApi::wait_and_acquire(HANDLE& mutex_handle, LPCWSTR mutex_name) noexcept
+{
+    if (locked(mutex_handle))
+        return 1;
+    locked(mutex_handle) = true;
+    return 0;
+}
+
+DWORD TestMutexApi::release(HANDLE& mutex_handle, LPCWSTR mutex_name) noexcept
+{
+    locked(mutex_handle) = false;
+    return 0;
+}
+
+// Testing Create and Destroy
+TEST(NamedMutexTests, CreateAndDestroy)
+{
+    auto& dbe = TestMutexApi::dummy_back_end;
+    std::list<TestMutexApi::dummy_mutex>::const_iterator it;
+    {
+        TestNamedMutex mutex(L"test-lifetime");
+        it = std::find(dbe.cbegin(), dbe.cend(), TestNamedMutex::mangle_name(L"test-lifetime"));
+        ASSERT_NE(it, dbe.cend());  // Added name to database -> Create called
+        ASSERT_EQ(it->refcount, 1); // destroy called once -> Create called once
+        ASSERT_FALSE(it->locked);   // Not locked -> wait_and_acquire not called
+
+        {
+            TestNamedMutex mutex_2(L"test-lifetime");
+            it = std::find(dbe.cbegin(), dbe.cend(), TestNamedMutex::mangle_name(L"test-lifetime"));
+            ASSERT_NE(it, dbe.cend());  // Name still in -> destroy not called
+            ASSERT_EQ(it->refcount, 2); // create called twice
+            ASSERT_FALSE(it->locked);   // Still not locked -> wait_and_acquire not called
+        }
+
+        it = std::find(dbe.cbegin(), dbe.cend(), TestNamedMutex::mangle_name(L"test-lifetime"));
+        ASSERT_NE(it, dbe.cend());  // Name not in database -> destroy called once
```

Comment:
If I understood the test correctly the comment needs some update, because the name is still in the "database".

Reply:
You're right. Fixed.